src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs

```diff
@@ -0,0 +1,302 @@
+        {
+            while (true)
+            {
+                _eventQueue.Add(() => RemoveExpiredValues());
+                await Task.Delay(interval).ConfigureAwait(false);
+            }
+        }
+
+        public void SetValue(TKey key, TValue value)
+        {
+            SetValue(key, value, DateTime.MaxValue);
+        }
+
+        public bool SetValue(TKey key, TValue value, DateTime expirationTime)
+        {
+            if (key == null)
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            if (value == null)
+                throw LogHelper.LogArgumentNullException(nameof(value));
+
+            // item already expired
+            if (expirationTime < DateTime.UtcNow)
+                return false;
+
+            // just need to update value and move it to the top
+            if (_map.ContainsKey(key))
+            {
+                var newCacheItem = new LRUCacheItem<TKey, TValue>(key, value, expirationTime);
+                // make sure node hasn't been removed by a different thread
+                if (_map.TryGetValue(key, out var cacheItem))
+                {
+                    _eventQueue.Add(() =>
+                    {
+                        _doubleLinkedList.Remove(cacheItem);
+                        _eventQueue.Add(() => _doubleLinkedList.AddFirst(newCacheItem));
+
+                    });
+                }
+                else
+                {
+                    _eventQueue.Add(() => _doubleLinkedList.AddFirst(newCacheItem));
+                }
+
+                return true;
+            }
+            else
+            {
+                // if cache is full, then remove the least recently used node
+                if (_map.Count >= _capacity)
+                {                
+                    _eventQueue.Add(() =>
+                    {
+                        var lru = _doubleLinkedList.Last;
+                        _map.TryRemove(lru.Value.Key, out _);
+                        _doubleLinkedList.Remove(lru);
+                    });
+                }
+
+                // add a new node
+                var node = new LRUCacheItem<TKey, TValue>(key, value, expirationTime);
+                _eventQueue.Add(() => _doubleLinkedList.AddFirst(node));
+                _map[key] = node;
+
+                return true;
+            }
+        }
+
+        /// Each time a node gets accessed, it gets moved to the beginning (head) of the list.
+        public bool TryGetValue(TKey key, out TValue value)
+        {
+            if (key == null)
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            if (!_map.ContainsKey(key))
+            {
+                value = default;
+                return false;
+            }
+
+            LRUCacheItem<TKey, TValue> cacheItem;
+            // make sure node hasn't been removed by a different thread
+            if (_map.TryGetValue(key, out cacheItem))
```

Comment:
cacheItem can be inline in the out statement, as is done above

Reply:
Thanks, fixed! https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1575/commits/38b55da469dcd67dfc96114d15fc2edf2a3802ba